🐛 Create schema: exception thrown generating error response

### Describe the bug.

I got an internal server error while passing invalid input to this endpoint

2023-11-01T17:33:43.255+00:00Copy[ERROR] AttributeError: 'DataProductSchema' object has no attribute 'error_traceback'Traceback (most recent call last):  File "/var/task/create_schema.py", line 179, in handler    error_msg = f"schema for {table_name} has failed validation with the following error: {schema.error_traceback}" | [ERROR] AttributeError: 'DataProductSchema' object has no attribute 'error_traceback' Traceback (most recent call last):   File "/var/task/create_schema.py", line 179, in handler     error_msg = f"schema for {table_name} has failed validation with the following error: {schema.error_traceback}"
-- | --




### To Reproduce

I sent a request to the create schema endpoint but forgot to wrap the JSON in a "schema" object.

i.e.

```json
{
  "tableDescription": "",
  "columns": [
    {
      "name": "id",
      "type": "bigint",
      "description": ""
    },
    {
      "name": "status",
      "type": "string",
      "description": ""
    },
    {
      "name": "noms_id",
      "type": "string",
      "description": ""
    },
    {
      "name": "first_name",
      "type": "string",
      "description": ""
    },
    {
      "name": "last_name",
      "type": "string",
      "description": ""
    },
    {
      "name": "date_of_birth",
      "type": "date",
      "description": ""
    },
    {
      "name": "pnc",
      "type": "string",
      "description": ""
    },
    {
      "name": "agy_loc_id",
      "type": "string",
      "description": ""
    },
    {
      "name": "created_date",
      "type": "date",
      "description": ""
    },
    {
      "name": "crn",
      "type": "string",
      "description": ""
    },
    {
      "name": "source_system",
      "type": "string",
      "description": ""
    },
    {
      "name": "ldu_code",
      "type": "string",
      "description": ""
    },
    {
      "name": "probation_area_code",
      "type": "string",
      "description": ""
    },
    {
      "name": "referral_system",
      "type": "string",
      "description": ""
    },
    {
      "name": "tact_type",
      "type": "string",
      "description": ""
    },
    {
      "name": "caseworker",
      "type": "string",
      "description": ""
    },
    {
      "name": "ct_officer",
      "type": "string",
      "description": ""
    },
    {
      "name": "immigration_team_key",
      "type": "string",
      "description": ""
    }
  ]
}
```

instead of `{"schema": {...}}`


### Expected Behaviour

A 400 response with the validation error

### Additional context

_No response_

--- a/containers/daap-create-schema/CHANGELOG.md
+++ b/containers/daap-create-schema/CHANGELOG.md
@@ -9,6 +9,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+## [1.1.1]
+
+### Changed
+
+- fix error response if "schema" key missing in POST
 - update Dockerfile COPY command fixing `LAMBDA_TASK_ROOT` typo
 
 ## [1.1.0]

--- a/containers/daap-create-schema/config.json
+++ b/containers/daap-create-schema/config.json
@@ -1,6 +1,6 @@
 {
     "name": "daap-create-schema",
-    "version": "1.1.0",
+    "version": "1.1.1",
     "registry": "ecr",
     "ecr": {
       "role": "arn:aws:iam::013433889002:role/modernisation-platform-oidc-cicd",

--- a/containers/daap-create-schema/src/var/task/create_schema.py
+++ b/containers/daap-create-schema/src/var/task/create_schema.py
@@ -112,11 +112,22 @@ def handler(event, context):
             HTTPStatus.BAD_REQUEST, event=event, message=error_message
         )
 
+    input_schema = request_body.get("schema")
+
+    if input_schema is None:
+        error_msg = (
+            "a 'schema' object was not passed in the request, "
+            "did you pass {table_schema} instead of {'schema': {table_schema}}?"
+        )
+        return format_error_response(
+            HTTPStatus.BAD_REQUEST, event=event, message=error_msg
+        )
+
     schema = DataProductSchema(
         data_product_name=data_product_name,
         table_name=table_name,
         logger=logger,
-        input_data=request_body.get("schema"),
+        input_data=input_schema,
     )
 
     if schema.exists:

--- a/containers/daap-create-schema/tests/unit/create_schema_test.py
+++ b/containers/daap-create-schema/tests/unit/create_schema_test.py
@@ -47,11 +47,22 @@ def test_s3_copy_folder_to_new_folder(s3_client):
 
 @pytest.mark.parametrize("method", [HTTPMethod.GET, HTTPMethod.PUT, HTTPMethod.DELETE])
 def test_http_method_fail(fake_event, fake_context, method):
+    with patch("create_schema.DataPlatformLogger"):
+        response = handler(event=fake_event, context=fake_context)
+
+        assert (
+            json.loads(response["body"])["error"]["message"]
+            == f"Sorry, {method} isn't allowed."
+        )
+
+
+@pytest.mark.parametrize("body_content", [{"TableDescription": "test_name"}])
+def test_schema_key_fail(fake_event, fake_context, body_content):
     response = handler(event=fake_event, context=fake_context)
 
-    assert (
-        json.loads(response["body"])["error"]["message"]
-        == f"Sorry, {method} isn't allowed."
+    assert json.loads(response["body"])["error"]["message"] == (
+        "a 'schema' object was not passed in the request, "
+        "did you pass {table_schema} instead of {'schema': {table_schema}}?"
     )
 
 

